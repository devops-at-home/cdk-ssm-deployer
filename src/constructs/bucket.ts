import { Construct } from 'constructs';
import { CfnOutput } from 'aws-cdk-lib';
import { Bucket, BucketEncryption, CfnBucket } from 'aws-cdk-lib/aws-s3';

export class BucketWithEventBridge extends Construct {
  public bucket: Bucket;
  constructor(scope: Construct, id: string) {
    super(scope, id);

    this.bucket = new Bucket(this, 'Bucket', {
      encryption: BucketEncryption.S3_MANAGED,
      versioned: true,
    });

    const cfnBucket = this.bucket.node.defaultChild as CfnBucket;
    cfnBucket.addPropertyOverride(
      'EventBridgeConfiguration.EventBridgeEnabled',
      'true'
    );

    new CfnOutput(this, 'bucketWebsiteUrl', {
      value: this.bucket.bucketName,
    });
  }
}
