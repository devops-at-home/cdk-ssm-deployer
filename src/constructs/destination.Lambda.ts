import { Handler, EventBridgeEvent, Context } from 'aws-lambda';

export const handler: Handler = async (
  event: EventBridgeEvent<any, any>,
  context: Context
) => {
  console.log(JSON.stringify(event));
  console.log(JSON.stringify(context));

  /* {
    currentVer: '',
    nextVer: {
      major: '',
      minor: '',
      patch: ''
    }
  }
  */
};
