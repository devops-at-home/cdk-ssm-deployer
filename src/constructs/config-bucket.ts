import { CfnOutput, Duration } from 'aws-cdk-lib';
import { Bucket, BucketEncryption, CfnBucket } from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

export class BucketWithEventBridge extends Construct {
  public bucket: Bucket;
  constructor(scope: Construct, id: string) {
    super(scope, id);

    this.bucket = new Bucket(this, 'Bucket', {
      encryption: BucketEncryption.S3_MANAGED,
      versioned: true,
      lifecycleRules: [
        {
          noncurrentVersionExpiration: Duration.days(60),
        },
      ],
    });

    const cfnBucket = this.bucket.node.defaultChild as CfnBucket;
    cfnBucket.addPropertyOverride(
      'NotificationConfiguration.EventBridgeConfiguration.EventBridgeEnabled',
      'true'
    );

    new CfnOutput(this, 'ConfigBucket', {
      value: this.bucket.bucketName,
    });
  }
}
