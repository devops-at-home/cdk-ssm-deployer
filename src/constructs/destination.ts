import { Construct } from 'constructs';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { Rule } from 'aws-cdk-lib/aws-events';
import { AwsApi } from 'aws-cdk-lib/aws-events-targets';
import { PolicyStatement } from 'aws-cdk-lib/aws-iam';

type SSMDestinationProps = {
  bucketName: string;
  destination: string;
};

export class SSMDestination extends Construct {
  constructor(scope: Construct, id: string, props: SSMDestinationProps) {
    super(scope, id);

    const { bucketName, destination } = props;

    new NodejsFunction(this, 'Lambda', {
      functionName: `deployer-${destination}-get-current-version`,
    });

    const s3BucketRule = new Rule(this, 'Rule', {
      ruleName: `deployTrigger-${destination}`,
      eventPattern: {
        source: ['aws.s3'],
        detail: {
          eventSource: ['s3.amazonaws.com'],
          eventName: ['PutObject'],
          requestParameters: {
            bucketName,
            key: `${destination}.zip`,
          },
        },
      },
    });

    const policyStatement = new PolicyStatement({
      actions: ['ssm:*'],
      resources: ['*'],
    });

    const ssmApiAction = new AwsApi({
      action: 'action',
      service: 'service',
      apiVersion: 'apiVersion',
      catchErrorPattern: 'catchErrorPattern',
      parameters: { destination },
      policyStatement,
    });

    s3BucketRule.addTarget(ssmApiAction);

    // TODO: OIDC permissions for git repo
    // - write to bucket folder (zip only)
    // - invoke version lambda (is it possible to restrict input params) -> only way to do this is a lambda per destination

    // TODO: ssm role and permissions to
    // - read from bucket folder
    // - invoke version lambda
    // - trigger deployment

    // TODO: custom resource to create destination folder in bucket
  }
}
