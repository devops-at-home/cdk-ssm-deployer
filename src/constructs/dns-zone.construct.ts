import { ResourceProps } from 'aws-cdk-lib';
import {
    ARecord,
    IPublicHostedZone,
    NsRecord,
    PublicHostedZone,
    RecordTarget,
} from 'aws-cdk-lib/aws-route53';
import { StringParameter } from 'aws-cdk-lib/aws-ssm';
import { Construct } from 'constructs';

interface DNSZoneProps extends ResourceProps {
    instanceName: string;
    environment: string;
    paramPrefix?: string;
}

export class DNSZone extends Construct {
    public readonly parentHostedZone: IPublicHostedZone;
    public readonly childHostedZone: IPublicHostedZone;

    constructor(scope: Construct, id: string, props: DNSZoneProps) {
        super(scope, id);

        const { instanceName, environment } = props;

        const paramPrefix = props.paramPrefix ?? '/edgeDevices';

        const parentZoneId = StringParameter.fromStringParameterName(
            this,
            'StringParamParentZoneId',
            `${paramPrefix}/infra/parentZoneId`
        ).stringValue;

        const parentZoneName = StringParameter.fromStringParameterName(
            this,
            'StringParamParentZoneName',
            `${paramPrefix}/infra/parentZoneName`
        ).stringValue;

        const tsIP = StringParameter.fromStringParameterName(
            this,
            'StringParamTsIP',
            `${paramPrefix}/instanceName/tsIP`
        ).stringValue;

        const zoneName = `${instanceName}.${
            environment === 'test' ? 'test.' : ''
        }${parentZoneName}`;

        this.childHostedZone = new PublicHostedZone(this, 'ChildPublicHostedZone', {
            zoneName,
        });

        this.parentHostedZone = PublicHostedZone.fromPublicHostedZoneAttributes(
            this,
            'PublicHostedZone',
            {
                zoneName: parentZoneName,
                hostedZoneId: parentZoneId,
            }
        );

        new NsRecord(this, 'ParentNSRecord', {
            values: this.childHostedZone.hostedZoneNameServers!,
            recordName: zoneName,
            zone: this.parentHostedZone,
        });

        new ARecord(this, 'ChildARecord', {
            target: RecordTarget.fromIpAddresses(tsIP),
            recordName: `*.${zoneName}`,
            zone: this.childHostedZone,
        });
    }
}
