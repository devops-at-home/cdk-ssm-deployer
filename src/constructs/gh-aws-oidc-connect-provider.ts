import { Construct } from 'constructs';
import { Aws } from 'aws-cdk-lib';
import { CfnOIDCProvider } from 'aws-cdk-lib/aws-iam';

// TODO: test https://github.com/askulkarni2/cdk-gh-aws-oidc-connect/blob/main/test/oidc-connect.test.ts
// FROM: https://github.com/askulkarni2/cdk-gh-aws-oidc-connect/blob/main/src/provider.ts
/**
 * Represents a GitHub OIDC provider.
 */
export interface IGitHubActionsOidcProvider {
  /**
   * The ARN of the OIDC provider.
   */
  readonly providerArn: string;
}

/**
 * Defines an OIDC provider for GitHub workflows.
 *
 * Please note that only a single instance of this provider can be installed in
 * an AWS account.
 */
export class GitHubActionsOidcProvider
  extends Construct
  implements IGitHubActionsOidcProvider
{
  /**
   * The OIDC domain for GitHub.
   */
  public static readonly DOMAIN = 'token.actions.githubusercontent.com';

  /**
   * The OIDC domain thumbprint for GitHub.
   */
  public static readonly THUMBPRINT =
    'a031c46782e6e6c662c2c87c76da9aa62ccabd8e';

  /**
   * @param account The AWS account for which you want to obtain the OIDC
   * provider. If not specified, we will use the current account.
   *
   * @returns The singleton GitHub OIDC provider for an account.
   */
  public static forAccount(account?: string): IGitHubActionsOidcProvider {
    return {
      providerArn: arnForAccount(account),
    };
  }

  /**
   * The ARN of the OIDC provider.
   */
  public readonly providerArn: string;

  constructor(scope: Construct, id: string, clientIdList: string[]) {
    super(scope, id);

    const provider = new CfnOIDCProvider(scope, `${id}.GithubOidcProvider`, {
      url: `https://${GitHubActionsOidcProvider.DOMAIN}`,
      thumbprintList: [GitHubActionsOidcProvider.THUMBPRINT],
      clientIdList,
    });

    this.providerArn = provider.attrArn;
  }
}

function arnForAccount(account?: string): string {
  account = account ?? Aws.ACCOUNT_ID;
  return `arn:aws:iam::${account}:oidc-provider/${GitHubActionsOidcProvider.DOMAIN}`;
}
