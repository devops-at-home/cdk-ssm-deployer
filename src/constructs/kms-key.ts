import { CfnOutput } from 'aws-cdk-lib';
import { Key } from 'aws-cdk-lib/aws-kms';
import { Construct } from 'constructs';

export class KmsKey extends Construct {
  public key: Key;
  constructor(scope: Construct, id: string) {
    super(scope, id);

    this.key = new Key(this, 'Key');

    new CfnOutput(this, 'Output', {
      value: this.key.keyId,
    });
  }
}
