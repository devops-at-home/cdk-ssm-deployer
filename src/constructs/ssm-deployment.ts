import { Construct } from 'constructs';
import { Rule } from 'aws-cdk-lib/aws-events';
import { AwsApi } from 'aws-cdk-lib/aws-events-targets';
import { PolicyStatement } from 'aws-cdk-lib/aws-iam';

type SSMDeploymentProps = {
  bucketName: string;
  destination: string;
};

export class SSMDeployment extends Construct {
  constructor(scope: Construct, id: string, props: SSMDeploymentProps) {
    super(scope, id);

    const { bucketName, destination } = props;

    const s3BucketRule = new Rule(this, 'Rule', {
      ruleName: `deployTrigger-${destination}`,
      eventPattern: {
        source: ['aws.s3'],
        detail: {
          eventSource: ['s3.amazonaws.com'],
          eventName: ['PutObject'],
          requestParameters: {
            bucketName,
            key: `${destination}.zip`,
          },
        },
      },
    });

    const policyStatement = new PolicyStatement({
      actions: ['ssm:*'],
      resources: ['*'],
    });

    // TODO: Invoke deployments documents to add and remove workloads when config added or removed from S3

    const ssmApiAction = new AwsApi({
      action: 'action',
      service: 'service',
      apiVersion: 'apiVersion',
      catchErrorPattern: 'catchErrorPattern',
      parameters: { destination },
      policyStatement,
    });

    s3BucketRule.addTarget(ssmApiAction);
  }
}
