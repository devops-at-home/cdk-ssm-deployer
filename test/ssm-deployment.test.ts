import { Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { SSMDeployment } from '../src/constructs/ssm-deployment';

describe('SSMDeployment', () => {
  const stack = new Stack();
  new SSMDeployment(stack, 'SSMDeployment', {
    bucketName: 'someName',
    destination: 'h6060-001',
  });

  const template = Template.fromStack(stack);

  test('Snapshot', () => {
    expect(template.toJSON()).toMatchSnapshot();
  });
});
