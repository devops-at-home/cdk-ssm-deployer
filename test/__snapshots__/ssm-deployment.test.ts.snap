// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SSMDeployment Snapshot 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "AWSb4cf1abd4e4f4bc699441af7ccd9ec371511E620": Object {
      "DependsOn": Array [
        "AWSb4cf1abd4e4f4bc699441af7ccd9ec37ServiceRoleDefaultPolicy4D43A7C1",
        "AWSb4cf1abd4e4f4bc699441af7ccd9ec37ServiceRole9FFE9C50",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "51cea70005434e28deb06e10b999a505b48f475c7a4762c082271c39c5a792d5.zip",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "AWSb4cf1abd4e4f4bc699441af7ccd9ec37ServiceRole9FFE9C50",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWSb4cf1abd4e4f4bc699441af7ccd9ec37ServiceRole9FFE9C50": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "AWSb4cf1abd4e4f4bc699441af7ccd9ec37ServiceRoleDefaultPolicy4D43A7C1": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "ssm:*",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "AWSb4cf1abd4e4f4bc699441af7ccd9ec37ServiceRoleDefaultPolicy4D43A7C1",
        "Roles": Array [
          Object {
            "Ref": "AWSb4cf1abd4e4f4bc699441af7ccd9ec37ServiceRole9FFE9C50",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "SSMDeploymentRule2C3EC80D": Object {
      "Properties": Object {
        "EventPattern": Object {
          "detail": Object {
            "eventName": Array [
              "PutObject",
            ],
            "eventSource": Array [
              "s3.amazonaws.com",
            ],
            "requestParameters": Object {
              "bucketName": "someName",
              "key": "h6060-001.zip",
            },
          },
          "source": Array [
            "aws.s3",
          ],
        },
        "Name": "deployTrigger-h6060-001",
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "AWSb4cf1abd4e4f4bc699441af7ccd9ec371511E620",
                "Arn",
              ],
            },
            "Id": "Target0",
            "Input": "{\\"service\\":\\"service\\",\\"action\\":\\"action\\",\\"parameters\\":{\\"destination\\":\\"h6060-001\\"},\\"catchErrorPattern\\":\\"catchErrorPattern\\",\\"apiVersion\\":\\"apiVersion\\"}",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "SSMDeploymentRuleAllowEventRuleSSMDeploymentRuleRuleTarget0HandlerAC0F46BD52D63CFF": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "AWSb4cf1abd4e4f4bc699441af7ccd9ec371511E620",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "SSMDeploymentRule2C3EC80D",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
